### ISSUE TITLE
Current master branch: mvn -Pprod yields broken app

### ISSUE THREAD

**Issue by erikkemperman** [jhipster/generator-jhipster#3017]
##### **Overview of the issue**

_Something_ about recent commit(s) causes the `build` gulp task to produce a broken app... I'm looking into it now but if anyone has a clue it'd be appreciated! At the moment, I suspect the refactor of angular scripts (wrapping in self-invoking function, use `<fn>.$inject`).

For what it's worth, all was fine with the master branch circa last Friday morning, CET.

The errors in console:
```sh
Uncaught TypeError: (intermediate value)(...) is not a function(anonymous function) @ translation.handler.js:40
 Uncaught Error: [$injector:unpr] Unknown provider: LanguageProvider <- Language <- stateHandler  @ angular.js:68
http://errors.angularjs.org/1.5.0/$injector/unpr?p0=LanguageProvider%20%3C-%20Language%20%3C-%20stateHandler
```

##### **Motivation for or Use Case** 

Trying to rebase my PR on current master, can't proceed now.

##### **JHipster Version(s)** 

Latest master branch.

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder** 

```js
{
  "generator-jhipster": {
    "jhipsterVersion": "<MASTER>",
    "baseName": "main",
    "packageName": "com.mycompany.myapp",
    "packageFolder": "com/mycompany/myapp",
    "authenticationType": "session",
    "hibernateCache": "ehcache",
    "clusteredHttpSession": "no",
    "websocket": "no",
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "prodDatabaseType": "mysql",
    "searchEngine": "no",
    "buildTool": "maven",
    "enableSocialSignIn": false,
    "useSass": false,
    "enableTranslation": true,
    "applicationType": "monolith",
    "testFrameworks": [
      "gatling",
      "protractor"
    ]
  }
}
```

##### **Browsers and Operating System** 

Tested in latest Chromium and Firefox on MacOSX.

##### **Reproduce the error** 

```sh
yo jhipster
mvn -Pprod
#open browser localhost:8080
```

**Comment by erikkemperman** [comment]
Found the problem... It's a classic.

If translation is enabled, `bower_components/messageformat/locale/{en,fr}.js` are injected into index.html. These contain lines like
```js
MessageFormat.locale.en=function(n){return n===1?"one":"other"}
```

Note the absence of `;` concluding the statement. 

When these are concatenated by `gulp usemin` with the angular scripts, this goes wrong because these were recently changed to follow the self-invoking function pattern,
```js
(function(){
    /*... */
})();
```

What happens in the concatenated script is that the round brackets of the first angular script after the locale statements are parsed as argument list, making an invocation of the locale function.

The proper but tedious fix would be to prepend all our angular scripts with a semi-colon, to terminate statements that precede it in concatenation. For the moment though, we could just make sure the locales are always injected last... Wanting to get on with what I was actually doing, I will make a PR for that quick fix shortly.

**Comment by erikkemperman** [comment]
By the way, I am not sure if I understand the rationale for introducing the self-invoking function pattern? 

To my knowledge this is usually done to create a scope for 'private' functions and variables, and/or to facilitate slightly more effective minification (mangling) due to this extra local scope.

But the few scripts I just looked at don't seem to have any local `var`s and mangling is currently turned off at the `uglify` plugin.

So if I'm missing some obvious benefit here, I'd love to learn about it!

**Comment by deepu105** [comment]
We are just following john papas angular style guides

I hate to have semicolon on files, may be we could do that during uglify
process?

**Comment by erikkemperman** [comment]
It's not pretty, I agree. The workaround I sent earlier isn't going to win any awards either.

The problem isn't uglify though, it's gulp-usemin, which does the concatenation but doesn't appear to allow passing separators like ';'.  

I'm all for style-guide adherence, but not if breaks my build :-) There are always valid exceptions, and I was just curious what the self-invoking functions accomplish here. If anything I'd guess it currently _adds_ a tiny bit to the minified size, what with all the extra `function` keywords!

**Comment by deepu105** [comment]
Shouldn't we try to PR messageformat to have ending semicolon?

**Comment by erikkemperman** [comment]
Another possibility would be to use the `negate_iife` option of `uglify`, which will transform the currently used pattern to something which, I seem to remember, fixes this kind of problem because it changes precedence. Would have to check that, though. Only in the minified output, so nothing awkward would have to be added in the source.

PRing messageformat... But we're getting a pretty old version of it indirectly, via [`bower-angular-translate-interpolation-messageformat`](https://github.com/angular-translate/bower-angular-translate-interpolation-messageformat/blob/2.9.2/bower.json). The latest releases don't have that `locale` directory at all.

**Comment by erikkemperman** [comment]
Cool, that `negate_iife` does the trick! Much cleaner and future-proof than my earlier workaround, so I'll amend my PR.

**Comment by erikkemperman** [comment]
Oops, celebrated too early. It's late here. Changed the PR back to workaround, and signing off for the day...

**Comment by deepu105** [comment]
Ha ha go home Erik you are drunk :P

**Comment by erikkemperman** [comment]
@deepu105 :-) Nope, can't even throw in that excuse! Just a long day, and slightly annoyed this got in the way of finalizing the webapp layout change (which I think is mostly done, so there's that).

**Comment by andidev** [comment]
@erikkemperman 
PR #2970 will move the locale scripts to the vendor block along with the other lib dependencies since they actually belong there. 
Is that okey? Or will that reintroduce this issue?

**Comment by erikkemperman** [comment]
Agree that is a better place for them, but yes it might have the same issue. It depends on what comes after the locale includes, so might get lucky... Still, I'd like a more robust, permanent solution for this problem, if possible?

**Comment by deepu105** [comment]
May be we can intercept the gulp stream right before concat and append,
prepend semicolon as required. Either using any plugin or a custom stream
processing task

**Comment by erikkemperman** [comment]
Like adding gulp-header to the usemin pipeline, just before concat, to add the semicolon.. That should work.

**Comment by deepu105** [comment]
closed via #3020
